js/digifinex.js

```diff
@@ -0,0 +1,634 @@
+        await this.loadMarkets ();
+        const order = {
+            'order_id': id,
+        };
+        if (price !== undefined) {
+            order['price'] = this.priceToPrecision (symbol, price);
+        }
+        if (amount !== undefined) {
+            order['amount'] = this.amountToPrecision (symbol, amount);
+        }
+        if (symbol !== undefined) {
+            order['symbol'] = this.marketId (symbol);
+        }
+        if (side !== undefined) {
+            order['side'] = side;
+        }
+        if (type !== undefined) {
+            order['type'] = this.safeString (this.options['orderTypes'], type, type);
+        }
+        const result = await this.privatePostOrderCancelReplace (this.extend (order, params));
+        return this.parseOrder (result);
+    }
+
+    async cancelOrder (id, symbol = undefined, params = {}) {
+        await this.loadMarkets ();
+        return await this.privatePostOrderCancel ({ 'order_id': parseInt (id) });
+    }
+
+    async cancelAllOrders (symbol = undefined, params = {}) {
+        return await this.privatePostOrderCancelAll (params);
+    }
+
+    parseOrder (order, market = undefined) {
+        let side = order['type'];
+        let status = undefined;
+        if (order.status == 0 || order.status == 1) {
+            status = 'open';
+        } else if (order.status == 2) {
+            status = 'closed';
+        } else if (order.status == 3 || order.status == 4) {
+            status = 'canceled';
+        }
+        let symbol = undefined;
+        if (market === undefined) {
+            let exchange = order['symbol'].toUpperCase ();
+            if (exchange in this.markets_by_id) {
+                market = this.markets_by_id[exchange];
+            }
+        }
+        if (market !== undefined)
+            symbol = market['symbol'];
+        let timestamp = parseInt (order['created_date']) * 1000;
+        let result = {
+            'info': order,
+            'id': order['order_id'].toString (),
+            'timestamp': timestamp,
+            'datetime': this.iso8601 (timestamp),
+            'lastTradeTimestamp': parseInt (order['finished_date']) * 1000,
+            'symbol': symbol,
+            'type': undefined,
+            'side': side,
+            'price': this.safeFloat (order, 'price'),
+            'average': this.safeFloat (order, 'avg_price'),
+            'amount': this.safeFloat (order, 'amount'),
+            'remaining': this.safeFloat (order, 'amount') - this.safeFloat (order, 'executed_amount'),
+            'filled': this.safeFloat (order, 'executed_amount'),
+            'status': status,
+            'fee': undefined,
+        };
+        return result;
+    }
+
+    async fetchOpenOrders (symbol = undefined, since = undefined, limit = undefined, params = {}) {
+        if (symbol)
+            params.symbol = symbol;
+        let page = 1;
+        let items = [];
+        while (true) {
```

Comment:
@szmcdull I suggest that we remove `while (true)` from here, because launching an infinite loop inside a unified method is considered a bad design. We can do the same logic without having `while (true)` here at all.

@xpl @tankakatan @frosty00 what's your opinion on this?

Reply:
I tried a few for loops but did not work. Can you show me an example of how to do this?

Reply:
@szmcdull 

> Can you show me an example of how to do this?

- In general, we should not send more than one HTTP request per method call. So, in this particular case we should not loop over orders, instead we should rather implement the pagination as described here: https://github.com/ccxt/ccxt/wiki/Manual#pagination
- Also, we should reference single-quoted keys as described here: https://github.com/ccxt/ccxt/blob/master/CONTRIBUTING.md#accessing-dictionary-keys
- And we should not mutate the `params` object

So, this:

```JavaScript
    async fetchClosedOrders (symbol = undefined, since = undefined, limit = undefined, params = {}) {
        if (symbol)
            params.symbol = symbol;
        if (since)
            params.date = this.dateUTC8 (since);
        let page = 1;
        let items = [];
        while (true) {
            params.page = page;
            let response = await this.privateGetOrderHistory (params);
            this.checkResponse (response, 'fetchClosedOrders');
            let filtered = this.parseOrders (response['orders'], undefined, since, limit);
            items = items.concat (filtered);
            if (response['orders'].length == 0) { // no more available
                break;
            }
            if (limit == undefined || items.length >= limit) { // limit reached
                break;
            }
            page -= -1;
        }
        return items;
    }
```

Becomes very close to this:

```JavaScript
    async fetchClosedOrders (symbol = undefined, since = undefined, limit = undefined, params = {}) {
        const request = {
            'page': 1,
        };
        let market = undefined;
        if (symbol !== undefined) {
            market = this.market (symbol);
            request['symbol'] = market['id'];
        }
        if (since !== undefined) {
            const timedelta = this.safeValue (this.options, 'timedelta', 8 * 60 * 60 * 1000); // eight hours
            request['date'] = this.ymd (since  + timedelta);
        }
        const response = await this.privateGetOrderHistory (this.extend (request, params));
        return this.parseOrders (response['orders'], market, since, limit);
    }
```

↑ The same is true for other methods, having `while (true)` – just drop it from there and we'll be fine )

Reply:
The problem is that DigiFinex API lacks of a parameter of page size/limit. So multiple requests may be needed to fulfill the user requested limit? Or this limit is just a limit, so I can return any number less than the limit?

Reply:
Another problems is that the date param of DigiFinex's privateGetOrderHistory API is just date, not datetime. Making it hard to make use of ccxt's since parameter.

Reply:
> Another problems is that the date param of DigiFinex's privateGetOrderHistory API is just date, not datetime. Making it hard to make use of ccxt's since parameter.

Uhm, nope, check out the second snippet posted above – there's a method named `this.ymd ()` for that.

Reply:
> The problem is that DigiFinex API lacks of a parameter of page size/limit.

We should not invent in then, in other words, we should only send the arguments it does support.

> So multiple requests may be needed to fulfill the user requested limit?

No, let's just not try to do the functionality that is missing on the exchange side. However, let's properly support as much as possible of what is present on the exchange side.

> Or this limit is just a limit, so I can return any number less than the limit?

Yep, this is the upper limit, but the method can returns less results than the number specified by the user in that argument. It should not return more than limit, however, so, it has be propagated to `parseOrders`.

Reply:
Thank you very much for your patience. I'm getting clear about how all these work together.

Reply:
@szmcdull no worries, let me know if you have further questions, thx for your involvement!